drivers/nec/display.cr

```diff
@@ -0,0 +1,299 @@
+require "placeos-driver/interface/powerable"
+require "placeos-driver/interface/muteable"
+require "placeos-driver/interface/switchable"
+
+class Nec::Display < PlaceOS::Driver
+  include Interface::Powerable
+  include Interface::AudioMuteable
+
+  enum Input
+    Vga         = 1
+    Rgbhv       = 2
+    Dvi         = 3
+    HdmiSet     = 4
+    Video1      = 5
+    Video2      = 6
+    Svideo      = 7
+    Tuner       = 9
+    Tv          = 10
+    Dvd1        = 12
+    Option      = 13
+    Dvd2        = 14
+    DisplayPort = 15
+    Hdmi        = 17
+    Hdmi2       = 18
+    Hdmi3       = 130
+    Usb         = 135
+  end
+  include PlaceOS::Driver::Interface::InputSelection(Input)
+
+  # Discovery Information
+  tcp_port 7142
+  descriptive_name "NEC Display"
+  generic_name :Display
+
+  DELIMITER = 0x0D_u8
+
+  def on_load
+    # Communication settings
+    queue.delay = 120.milliseconds
+    transport.tokenizer = Tokenizer.new(Bytes[DELIMITER])
+    on_update
+  end
+
+  def on_update
+  end
+
+  def connected
+    schedule.every(50.seconds, true) do
+      do_poll
+    end
+  end
+
+  def disconnected
+    schedule.clear
+  end
+
+  def power(state : Bool)
+    # Do nothing if already in desired state
+    return if self[:power]? == state
+
+    if state
+      logger.debug { "-- NEC LCD, requested to power on" }
+      # 1 = Power On
+      data = MsgType::Command.build(Command::SetPower, 1)
+      send(data, name: "power", delay: 5.seconds)
+    else
+      logger.debug { "-- NEC LCD, requested to power off" }
+      # 4 = Power Off
+      data = MsgType::Command.build(Command::SetPower, 4)
+      send(data, name: "power", delay: 10.seconds, timeout: 10.seconds)
+    end
+  end
+
+  def power?(**options) : Bool
+    data = MsgType::Command.build(Command::PowerQuery)
+    send(data, **options, name: "power?").get
+    self[:power].as_bool
+  end
+
+  def switch_to(input : Input)
+    logger.debug { "-- NEC LCD, requested to switch to: #{input}" }
+    data = MsgType::SetParameter.build(Command::VideoInput, input.value)
+    send(data, name: "input", delay: 6.seconds)
+  end
+
+  enum Audio
+    Audio1      = 1
+    Audio2      = 2
+    Audio3      = 3
+    Hdmi        = 4
+    Tv          = 6
+    DisplayPort = 7
+  end
+
+  def switch_audio(input : Audio)
+    logger.debug { "-- NEC LCD, requested to switch audio to: #{input}" }
+    data = MsgType::SetParameter.build(Command::AudioInput, input.value)
+    send(data, name: "audio")
+  end
+
+  def auto_adjust
+    data = MsgType::SetParameter.build(Command::AutoSetup, 1)
+    send(data, name: "auto_adjust")
+  end
+
+  def brightness(val : Int32)
+    data = MsgType::SetParameter.build(Command::BrightnessStatus, val.clamp(0, 100))
+    send(data, name: "brightness")
+    send(MsgType::Command.build(Command::Save), name: "brightness_save") # Save the settings
```

Comment:
The save command looks to be the same regardless of what state you're trying to persist. With this in mind, to remove redundancy they can all be named the same and set to a low priority. That way if a set of settings are applied at once, `Command::Save` will only be sent once after all of these have been applied.

If you do this, just make sure the the result from the `send` for the actual setting change is returned, otherwise callers of the methods will be returned an error if their save command is overwritten (which is desired).